README
NAME
    Tree::Node - Memory-efficient tree nodes in Perl

REQUIREMENTS
    Perl 5.6.0 or newer is required. Only core modules are used.

    A C compiler to is required to build the module. (There is no Pure-perl
    version because this package was written to overcome limitations of
    Perl. See the DESCRIPTION section below.)

INSTALLATION
    Installation can be done using the traditional Makefile.PL method:

      perl Makefile.PL
      make
      make test
      make install

    (On Windows platforms you should use `nmake' instead.)

SYNOPSIS
      use Tree::Node;

      $node = Tree::Node->new(2);

      $node->set_child(0, $left);
      $node->set_child(1, $right);

      while ($node->key_cmp($key) < 0) {
        $node = $node->get_child(0);
      }

DESCRIPTION
    This module implements a memory-efficient node type (for trees, skip
    lists and similar data structures) for Perl.

    You may ask "Why bother implementing an ordered structure such as a tree
    when Perl has hashes built-in?" Since Perl is optimized for speed over
    memory usage, hashes (and lists) use a lot of memory.

    Using Devel::Size for a reference, a list with four elements
    (corresponding to a key, value, and two child node pointers) will use at
    least 120 bytes. A hash with four key/value pairs will use at least 228
    bytes. But an equivalent Tree::Node object will use at least 68 bytes.
    (However, see the KNOWN ISSUES section below for caveats regarding
    memory usage.)

    So the purpose of this package is to provide a simple low-level Node
    class which can be used as a base class to implement various kinds of
    tree structures. Each node has a key/value pair and a variable number of
    "children" pointers.

    How nodes are organized or the algorithm used to organize them is for
    you to implement.

REVISION HISTORY
    The following changes have been made since the last release:

    0.09
                - Bumped up versuin for upload to GitHub.

    See the Changes file for a more detailed revision history.

KNOWN ISSUES
    This module implements a Perl wrapper around a C struct, which for the
    object-oriented inferface involves a blessed reference to a pointer to
    the struct. This overhead of about 45 bytes may make up for any memory
    savings that the C-based struct provided!

    So if you what you are doing is implementing a simple key/value lookup,
    then you may be better off sticking with hashes. If what you are doing
    requires a special structure that cannot be satisfied with hashes (even
    sorted hashes), or requires a very large number of nodes, then this
    module may be useful to you.

    Another alternative is to use the Procedural Interface.

SEE ALSO
    Tree::DAG_Node is written in pure Perl, but it offers a more flexible
    interface.

AUTHOR
    Robert Rothenberg <rrwo at cpan.org>

  Suggestions and Bug Reporting
    Feedback is always welcome. Please use the CPAN Request Tracker at
    http://rt.cpan.org to submit bug reports.

LICENSE
    Copyright (c) 2005,2007 Robert Rothenberg. All rights reserved. This
    program is free software; you can redistribute it and/or modify it under
    the same terms as Perl itself.



Changes
Revision history for Perl extension Tree::Node.

0.10

  [Possible Incompatabile Changes]
  - Calling the key_cmp method with undef dies.

  - Removed exprimental procedural interface.

  [Enhancements]
  - The set_key, force_set_key, set_value, set_child, and
    add_children methods are chainable.

  - Added Tree::Node::Clonable to support serialization.

  [Other Changes]
  - Renamed get_child_or_undef to get_child_maybe. Original
    method exists as an alias, but is deprecated.

  - Corrected typos in Changes

  - Reformatted Changes to CPAN::Changes::Spec

  - Removed Build.PL.

  - Misc cleanup of git repository.

  - Moved developer tests to xt directory.

  - Added more tests.

0.09  2011-06-19

  - Bumped up version for upload to GitHub.

0.08  2007-07-20

  - Minor tweaks and typo corrections.

0.07  2007-07-13

  - Distro is no longer signed (due to Module:Signature issues).

  - Fixed error with reading undefined key or value.

  - Added force_set_key mode to change key.

  - Test for leaks in procedural interface.

  - Improved pointer cast.

  - Minor changes to Build.PL

0.06  2005-07-28

  - Added Build.PL.

  - Added additional tests ("kwalitee"-related).

  - Uses Perl malloc macros instead of system malloc.

  - Updated documentation regarding memory usage.

  - Added exports of utility functions.

  - Added experimental procedural interface.

  - Added to_p_node method to get pointer.

0.051  2005-07-22

  - Fixed bad SIGNATURE file.

0.05  2007-07-22

  - Removed _rotate_children method.

  - Renamed SIZE macro to NODESIZE.

  - Fixed bug where node size allocation was too small!

  - Renamed _increment_child_count to add_children.

  - Add_children accepts a list of children to add (so it can add
    multiple children).

  - Added get_children method to return an array of children.

  - Readonly flag is set (this was disabled in v0.04).

  - Documentation updated accordingly.

  - Added developer tests.

0.04  2005-07-15

  - Added _increment_child_count and _rotate_children methods.

  - Updated documentation.

  - Removed XSLoader from dependencies, since CPANPLUS
    complains.

0.03  2005-07-02

  - Fixed issue with SIGNATURE file.

  - More updates to documentation.

0.02  2005-06-30

  - Minor updates to documentation.

  - Fixed missing include in TreeNode.c.

0.01  2005-06-29

  - Branched off from experimental Algorithm::SkipList::Node.



TreeNode.c

#include <stdlib.h>

#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"
#include "ppport.h"

#include "TreeNode.h"

Node * new(int child_count)
{
  Node * n;

  if ((child_count < 1) || (child_count > MAX_LEVEL))
    croak("child_count out of bounds: must be between [1..%d]", MAX_LEVEL);

#if USE_MALLOC
  n = malloc((size_t) NODESIZE(child_count));
  if (n == NULL)
    croak("unable to a allocate memory");
#else
  Newc(1, n, NODESIZE(child_count), char, Node);
#endif

  n->child_count  = child_count;

  n->key   = &PL_sv_undef;
  n->value = &PL_sv_undef;

  return n;
}

void DESTROY(Node* n)
{
  SvREFCNT_dec(n->key);
  SvREFCNT_dec(n->value);

#if USE_MALLOC
  free(n);
#else
  Safefree(n);
#endif
}

int child_count(Node * n)
{
  return n->child_count;
}

SV* get_child(Node * n, int index)
{
  if ((index >= n->child_count) || (index < 0))
    croak("index out of bounds: must be between [0..%d]", n->child_count-1);

  return SvREFCNT_inc(n->next[index]);
}

SV* get_child_maybe(Node * n, int index)
{
  if ((index >= n->child_count) || (index < 0))
    return &PL_sv_undef;
  else
    return SvREFCNT_inc(n->next[index]);
}

void set_child(Node* n, int index, SV* t)
{
  if ((index >= n->child_count) || (index < 0))
    croak("index out of bounds: must be between [0..%d]", n->child_count-1);

  if (SvOK(n->next[index]))
    sv_setsv(n->next[index], t);
  else
    n->next[index] = newSVsv(t);

}

void set_key(Node *n, SV* k)
{
  if (SvOK(n->key)) {
    croak("key is already set");
  }
  else {
    n->key = newRV_inc(k);
  }
}

void force_set_key(Node *n, SV* k)
{
  if (SvOK(n->key)) {
    warn("key is already set");
  }
  n->key = newRV_inc(k);
}

SV* get_key(Node *n)
{
  if (SvOK(n->key)) {
    return SvREFCNT_inc(SvRV(n->key));
  }
  else {
    return &PL_sv_undef;
  }
}

I32 key_cmp(Node* n, SV* k)
{
  if (SvOK(n->key)) {
    return sv_cmp(SvRV(n->key), k);
  }
  else {
    return -1;
  }
}

void set_value(Node *n, SV* v)
{
  n->value = newRV_inc(v);
}

SV* get_value(Node *n)
{
  if (SvOK(n->key)) {
    return SvREFCNT_inc(SvRV(n->value));
  }
  else {
    return &PL_sv_undef;
  }
}

int _allocated(Node* n)
{
  return NODESIZE(n->child_count);
}


Makefile.PL
use v5.10.1;

use ExtUtils::MakeMaker;

WriteMakefile(
    NAME           => 'Tree::Node',
    VERSION_FROM   => 'lib/Tree/Node.pm',
    PREREQ_PM      => {
      'parent'         => 0,
      'Carp'           => 0,
      'File::Copy'     => 0,
      'File::Spec'     => 0,
      'Scalar::Util'   => 0,
      'Test::Most'     => 0,
      'Test::Warnings' => 0,
    },
    ABSTRACT_FROM  => 'lib/Tree/Node.pm', # retrieve abstract from module
    AUTHOR         => 'Robert Rothenberg <rrwo at cpan.org>',
    LIBS           => [''],  # e.g., '-lm'
    DEFINE         => '',    # e.g., '-DHAVE_SOMETHING'
    INC            => '-I.', # e.g., '-I. -I/usr/include/other'
    OBJECT         => '$(O_FILES)',
);
if  (eval {require ExtUtils::Constant; 1}) {
  my @names = (qw());
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Tree::Node',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}


lib/Tree/Node/Clonable.pm
package Tree::Node::Clonable;

use strict;
use warnings;

use parent 'Tree::Node';

our $VERSION = '0.10';

use Carp;
use Scalar::Util qw/ blessed refaddr /;

=head1 NAME

Tree::Node::Clonable - a clonable Tree::Node

=head1 SYNOPSIS

  use Tree::Node::Clonable;

  my $node = Tree::Node::Clonable->new(10);

  ...

  my $data = $node->serialize();

  ...

  $node = Tree::Node::Clonable->deserialize($data);

=head1 DESCRIPTION

This class provides a serializable form of L<Tree::Node>.

=head1 METHODS

=head2 serialize

  $data = $node->serialize();

Returns a hash reference of the object.

It assumes the children are L<Tree::Node::Clonable> objects.

=cut

sub serialize {
    my ($self, $seen) = @_;

    return unless $self && blessed($self) && $self->isa(__PACKAGE__);

    $seen //= {};

    my $serialized = ( $seen->{refaddr $self} //=
      {
       key      => $self->key,
       value    => $self->value,
       size     => $self->child_count,
       children => [ map { serialize($_, $seen) } $self->get_children ],
      }
    );

    return $serialized;
}

=head2 deserialize

  $node = Tree::Node::Clonable->deserialize( \%data, @children );

Deserializes a serialized node.

=cut

sub deserialize {
    my ($class, $data, $seen) = @_;

    croak "cannot deserialize"
      unless (
              $data &&
              ref($data) eq 'HASH' &&
              exists $data->{key} &&
              exists $data->{value} &&
              exists $data->{children} &&
              ref($data->{children}) eq 'ARRAY'
             );

    $seen //= { };

    return $seen->{refaddr $data}
      if $seen->{refaddr $data};

    my $size = $data->{size} // ( 1 + $#{$data->{children}} );
    my $self = $class->new($size);

    $self->set_key($data->{key});
    $self->set_value($data->{value});

    for(my $i = 0; $i < $size; $i++) {
      if (my $child = $data->{children}->[$i]) {
        $self->set_child($i, $class->deserialize($child, $seen));
      }
    }

    $seen->{refaddr $data} = $self;

    return $self;
}

=head1 AUTHOR

Robert Rothenberg <rrwo at cpan.org>

=head2 Suggestions and Bug Reporting

Feedback is always welcome.  Please use the CPAN Request Tracker at
L<http://rt.cpan.org> to submit bug reports.

=head1 LICENSE

Copyright (c) 2005-2014 Robert Rothenberg. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut

1;


lib/Tree/Node.pm
package Tree::Node;

use 5.006;
use strict;
use warnings;

our $VERSION = '0.10';

require XSLoader;
XSLoader::load('Tree::Node', $VERSION);

use Exporter::Lite;

our @EXPORT = (qw/ MAX_LEVEL /);

1;
__END__

=head1 NAME

Tree::Node - Memory-efficient tree nodes in Perl

=begin readme,install

=head1 REQUIREMENTS

Perl 5.10.0 or newer is required.

A C compiler to is required to build the module.  (There is no Pure-perl
version because this package was written to overcome limitations of Perl.
See the L</DESCRIPTION> section below.)

=head1 INSTALLATION

Installation can be done using the traditional F<Makefile.PL> method:

  perl Makefile.PL
  make
  make test
  make install

(On Windows platforms you should use C<nmake> instead.)

=end readme,install

=for install stop

=head1 SYNOPSIS

  use Tree::Node;

  $node = Tree::Node->new(2);

  $node->set_child(0, $left);
  $node->set_child(1, $right);

  while ($node->key_cmp($key) < 0) {
    $node = $node->get_child(0);
  }

=head1 DESCRIPTION

This module implements a memory-efficient node type (for trees,
skip lists and similar data structures) for Perl.

You may ask "Why bother implementing an ordered structure such
as a tree when Perl has hashes built-in?"  Since Perl is optimized
for speed over memory usage, hashes (and lists) use a lot of memory.

Using L<Devel::Size> for a reference, a list with four elements
(corresponding to a key, value, and two child node pointers) will
use at least 120 bytes.  A hash with four key/value pairs will
use at least 228 bytes.  But an equivalent L<Tree::Node> object
will use at least 68 bytes.  (However, see the L</KNOWN ISSUES>
section below for caveats regarding memory usage.)

So the purpose of this package is to provide a simple low-level Node
class which can be used as a base class to implement various kinds
of tree structures.  Each node has a key/value pair and a variable
number of "children" pointers.

How nodes are organized or the algorithm used to organize them is
for you to implement.

=for readme stop

=head2 Object Oritented Interface

=over

=item new

  $node = Tree::Node->new( $child_count );

Creates a new node with C<$child_count> children.  Only as much space as is
needed is allocated, and the number of children cannot be expanded later
on.

C<$child_count> cannot exceed L</MAX_LEVEL>.

=item child_count

  $child_count = $node->child_count;

Returns the number of childen allocated.

=item set_key

  $node->set_key($key);

Sets the key. Once it is set, it cannot be changed.

=item force_set_key

  $node->force_set_key($key);

Sets the key, irregardless of whether it is already set. (Note
that many data structures assume that the key is never changed,
so you should only use this for cases where it is safe to do
so.)

=item key

  $key = $node->key;

Returns the node key.

=item key_cmp

  if ($node->key_cmp($key) < 0) { ... }

Compares the node key with a key using the Perl string comparison
function C<cmp>.  Returns -1 if the node key is less than the
argument, 0 if it is equal, and 1 if it is greater.

If the key is undefined, then it will always return -1 (even if the
argument is undefined).

This method can be overriden if you need a different comparison
routine.  To use numeric keys, for example:

  package Tree::Node::Numeric;

  use base 'Tree::Node';

  sub key_cmp {
    my $self = shift;
    my $key  = shift;
    return ($self->key <=> $key);
  }

=item set_value

  $node->set_value($value);

Sets the value of the node.  The value can be changed.

=item value

  $value = $node->value;

Returns the value of the node.

=item set_child

  $node->set_child($index, $child);

Sets the child node.  C<$index> must be between 0 and L</child_count>-1.
Dues when the C<$index> is out of bounds.

=item get_child

  $child = $node->get_child($index);

Returns the child node.  Dies when the C<$index> is out of bounds.

=item get_child_or_undef

  $child = $node->get_child_or_undef($index);

Like L</get_child>, but returns C<undef> rather than dying when the
C<$index> is out of bounds.

=item get_children

  @children = $node->get_children;

=item add_children

  $node->add_children(@children)

Increases the L</child_count> and allocates space for the child nodes
specified.  (The child nodes can be C<undef>.)

=item add_children_left

Same as L</add_children>, except that the new nodes are added to
the beginning rather than end of the node list.

=item MAX_LEVEL


Returns the maximum number of children. Defaults to the C constant
C<UCHAR_MAX>, which is usually 255.

=item _allocated

  $size = $node->_allocated;

This is a utility routine which says how much space is allocated for a
node.  It does not include the Perl overhead (see L</KNOWN ISSUES> below).

=item _allocated_by_child_count

  use Tree::Node ':utility';

  ...

  $size = _allocated_by_child_count( $child_count );

This is a utility routine which returns the amount of space that would be
allocated for a node with C<$child_count> children.

=back

=head2 Procedural Inferface

The experimental procedural interface was added in version 0.06 and
removed in version 0.10.

=for readme continue

=begin readme

=head1 REVISION HISTORY

The following changes have been made since the last release:

=for readme include file="Changes" type="text" start="^0.10" stop="^0.08"

See the F<Changes> file for a more detailed revision history.

=end readme

=for readme continue

=head1 KNOWN ISSUES

This module implements a Perl wrapper around a C struct, which for the
object-oriented inferface involves a blessed reference to a pointer to
the struct.  This overhead of about 45 bytes may make up for any memory
savings that the C-based struct provided!

So if you what you are doing is implementing a simple key/value lookup,
then you may be better off sticking with hashes.  If what you are doing
requires a special structure that cannot be satisfied with hashes (even
sorted hashes), or requires a very large number of nodes, then this module
may be useful to you.

=for readme stop

Packages such as L<Clone> and L<Storable> cannot properly handle Tree::Node
objects.

L<Devel::Size> may not properly determine the size of a node. Use the
L</_allocated> method to determine how much space is allocated for the
node in C.  This does not include the overhead for Perl to maintain a
reference to the C struct.

=for readme,install continue

=head1 SEE ALSO

L<Tree::DAG_Node> is written in pure Perl, but it offers a more
flexible interface.

=head1 AUTHOR

Robert Rothenberg <rrwo at cpan.org>

=head2 Suggestions and Bug Reporting

Feedback is always welcome.  Please use the CPAN Request Tracker at
L<http://rt.cpan.org> to submit bug reports.

=head1 LICENSE

Copyright (c) 2005-2014 Robert Rothenberg. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut

